### PR TITLE
feat: Enable TotalBytes metrics

### LINKED ISSUE #226: Total number of bytes processed metric
This would be a useful metric for:

* How well is an org adopting Argo Dataflow?
* What is the volume of data it is processing?

See #201

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>
closes #226

### CODE REVIEW

File: api/v1alpha1/metrics.go [hunk 1]
```diff
@@ -5,8 +5,9 @@ import (
...
 type Metrics struct {
-	Total   uint64            `json:"total,omitempty" protobuf:"varint,1,opt,name=total"`
-	Errors  uint64            `json:"errors,omitempty" protobuf:"varint,2,opt,name=errors"`
-	Rate    resource.Quantity `json:"rate,omitempty" protobuf:"bytes,3,opt,name=rate"` // current rate of messages per second
-	Retries uint64            `json:"retries,omitempty" protobuf:"bytes,4,opt,name=retries"`
+	Total      uint64            `json:"total,omitempty" protobuf:"varint,1,opt,name=total"`
+	Errors     uint64            `json:"errors,omitempty" protobuf:"varint,2,opt,name=errors"`
+	Rate       resource.Quantity `json:"rate,omitempty" protobuf:"bytes,3,opt,name=rate"` // current rate of messages per second
+	Retries    uint64            `json:"retries,omitempty" protobuf:"bytes,4,opt,name=retries"`
+	TotalBytes uint64            `json:"totalBytes,omitempty" protobuf:"varint,5,opt,name=totalBytes"`
```
Comment (sarabala1979) [inline api/v1alpha1/metrics.go:12]: Bytes will reach uint64 max soon. @alexec  Can we have TotalKiloBytes or TotalMBytes?

File: api/v1alpha1/source_statuses.go [hunk 2]
```diff
@@ -92,3 +92,22 @@ func (in SourceStatuses) IncrRetries(name string, replica int) {
...
+
+func (in SourceStatuses) GetTotalBytes() uint64 {
+	var v uint64
+	for _, s := range in {
+		v += s.GetTotalBytes()
+	}
+	return v
+}
+
+func (in SourceStatuses) IncrTotalBytes(name string, replica int, bytes uint64) {
```
Comment (alexec) [inline api/v1alpha1/source_statuses.go:104]: I think you can combine this func with `IncrTotal`

File: api/v1alpha1/metrics.go (hunk 1 above)
Comment (alexec) [reply api/v1alpha1/metrics.go:12]: try `resource.Quantity`

File: api/v1alpha1/source_statuses.go (hunk 2 above)
Comment (sarabala1979) [reply api/v1alpha1/source_statuses.go]: done

File: api/v1alpha1/metrics.go (hunk 1 above)
Comment (sarabala1979) [reply api/v1alpha1/metrics.go:12]: I will keep uint64. resource.Quantity has int64

Comment (alexec) [reply api/v1alpha1/metrics.go:12]: is `uint64` big enough do you think?  18446744073709551615 in max uint64

image
image

@whynowy do you want to add your thoughts?

File: runner/sidecar/sources.go [hunk 3]
```diff
@@ -168,4 +168,15 @@ func newSourceMetrics(source dfv1.Source, sourceName string) {
 		defer mu.RUnlock()
 		return float64(step.Status.SourceStatuses.Get(sourceName).GetRetries())
 	})
+
+	promauto.NewCounterFunc(prometheus.CounterOpts{
```
Comment (alexec) [inline runner/sidecar/sources.go:172]: can you update `metrics_test.go`?

**Review by alexec — CHANGES_REQUESTED** [review]
see comment, also need to run `make codegen`

File: test/e2e/metrics_test.go [hunk 4]
```diff
@@ -49,4 +49,7 @@ func TestMetrics(t *testing.T) {
 	ExpectMetric("sources_pending", 0)
 	ExpectMetric("sources_total", 1)
 	ExpectMetric("sources_retries", 0)
+	ExpectMetric("sources_totalBytes", 6)
+	SendMessageViaHTTP("my-msg")
+	ExpectMetric("sources_totalBytes", 12)
```
Comment (alexec) [inline test/e2e/metrics_test.go:54]: nice

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Can you please submit a PR to show this in the UI please?

**Pull request merged**